src/OmniSharp.Roslyn/OmniSharpWorkspace.cs

```diff
@@ -80,9 +85,51 @@ public void RemoveMetadataReference(ProjectId projectId, MetadataReference metad
 
         public void AddDocument(DocumentInfo documentInfo)
         {
+            var documentId = GetDocumentId(documentInfo.FilePath);
+            if (documentId != null && _miscellanousFiles.Contains(documentId))
+            {
+                //if the file has already been added as a misc file,
+                //because of a possible race condition between the updates of the project systems,
+                //remove the misc file and add it the document as required
+                _miscellanousFiles.Remove(documentId);
+                RemoveDocument(documentId);
+            }
+           
             OnDocumentAdded(documentInfo);
         }
 
+        public DocumentId AddMiscellaneousFileDocument(string filePath, string language)
+        {
+            if (MiscFilesProjectInfo == null)
```

Comment:
Yeah, move the check if the doc is already a regular document into this method :)